### ISSUE TITLE
Better error message when failing to parse array length

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/syn#28]
Rust allows arbitrary code as an array length:

``` rust
pub struct Screen(pub [Color; {
    fn holy_smokes() {
        println!("why am I part of this type?");
    }
    12288
}]);
```

To minimize compile time, syn supports integer literals only (EDIT: now a few other simple expressions too, but still limited). The difference in compile time is more than a factor of 2 between supporting integers only vs supporting expressions, so currently I believe the tradeoff makes sense.

Given that this is more restrictive than Rust, it would be nice to give a better message than the usual "failed to parse macro input" when the failure is related to an array length.

**Comment by dtolnay** [comment]
Triaging as out of scope for 0.12 but a major focus for 0.13.

**Comment by dtolnay** [comment]
Fixed in 6e1e505ea5a3f95d838666ed6391af1993266aa0.

```console
error: unsupported expression; enable syn's features=["full"]
  --> src/main.rs:8:31
   |
8  |   pub struct Screen(pub [Color; {
   |  _______________________________^
9  | |     fn holy_smokes() {
10 | |         println!("why am I part of this type?");
11 | |     }
12 | |     12288
13 | | }]);
   | |_^
```
